src/extension.ts

```diff
@@ -1,98 +1,65 @@
-import { jsLangs } from './util'
+import { parseVariables } from './utils'
 
 export const activate = () => {
-    // registerExtensionCommand('runWithSideCode', async () => {
-    //     const activeEditor = vscode.window.activeTextEditor
-    //     if (!activeEditor || activeEditor.viewColumn === undefined) return
-    //     await vscode.window.showTextDocument(vscode.Uri.parse(`${documentScheme}:${activeEditor.document.uri.toString(true)}`), {
-    //         preview: false,
-    //         viewColumn: vscode.ViewColumn.Beside,
-    //     })
-    // })
     type FsPath = string
     const activeTerminals = new Map<FsPath, vscode.Terminal>()
 
-    const checkRunButton = (textEditor: vscode.TextEditor | undefined): void => {
-        if (!textEditor || textEditor.viewColumn === undefined) return
-        let hasExec = !!getExtensionSetting('defaultExec')
-        if (!hasExec) hasExec = !!getExec(textEditor.document.languageId)
-        void vscode.commands.executeCommand('setContext', `terminal-code-runner.runBututon`, hasExec)
+    const checkDisplayRunButton = (textEditor: vscode.TextEditor | undefined): void => {
+        void vscode.commands.executeCommand('setContext', `terminal-code-runner.runButton`, getHasExec(textEditor))
     }
 
-    vscode.window.onDidChangeActiveTextEditor(checkRunButton)
-    checkRunButton(vscode.window.activeTextEditor)
+    vscode.window.onDidChangeActiveTextEditor(checkDisplayRunButton)
+    checkDisplayRunButton(vscode.window.activeTextEditor)
 
     registerExtensionCommand('runFile', async () => {
         const activeEditor = vscode.window.activeTextEditor
         if (!activeEditor || activeEditor.viewColumn === undefined) return
+
+        let exec = getExec(activeEditor)
+        if (!exec) {
+            void vscode.window.showWarningMessage(`No matched exec command!`)
+            return
+        }
+
         const { document } = activeEditor
         const { fsPath } = document.uri
         const fileDir = dirname(fsPath)
         const fileName = basename(fsPath)
-        const workspaceRoot = vscode.workspace.getWorkspaceFolder(document.uri)
-        const exec = (getExec(document.languageId) ?? getExtensionSetting('defaultExec'))
-            ?.replace(/\$workspaceRoot/g, workspaceRoot?.uri.fsPath ?? '')
-            .replace(/\$fileName/, fileName)
-            .replace(/\$dir/, fileDir)
-            .replace(/\$path/, fsPath)
-        if (!exec) {
-            void vscode.window.showWarningMessage(`No exec command for language ${document.languageId} is set`)
-            return
-        }
+        exec = await parseVariables(exec, document.uri)
 
-        const writeEmitter = new vscode.EventEmitter<string>()
-        const line = ''
+        // Uses different terminals for each file
+        const isDifferentTerminalsForEachFile = getExtensionSetting('differentTerminalsForEachFile')
+        const terminalKey = isDifferentTerminalsForEachFile ? fsPath : '--Reused Terminal--'
```

Comment:
```suggestion
        const terminalKey = isDifferentTerminalsForEachFile ? fsPath : 'Terminal Code Runner'
```

Also would you mind extending to union instead? eg this setting: `executeInTerminal: "alwaysActive" | "shared" | "file"`. I'm sorry for so much requested changes, I can handle this. I just want to use `alwaysActive`, but my friend `dedicated`, while many other users as I heard still prefer old behavior.

Reply:
Btw vscode tasks use a similar concept (`vscode.TaskPresentationOptions['panel']` type). You might be aware of global user tasks vscode feature, which uses it.

Reply:
Ok, you can do it yourself. As for the extension icon, you can also make it your self, or just keep lost.

> Maybe we can let ai generate a extension icon

I'm not familiar with ai draw, but recently it is very hot though, I just make a kidding....